### PR TITLE
Deprecate some versions of DoFTools::extract_dofs().

### PR DESCRIPTION
**Pull request by bangerth**
Ever since we started using C++11, it has bothered me that we have a lot of functions that return their results in output arguments (typically in last position, but not always -- e.g., if there are additional defaulted arguments). This can now be done more efficiently because we can move objects, and I've wanted to address this for at least some of the functions where this is the case.

Here is a start: A couple of functions in `DoFTools`.

### CODE REVIEW

File: source/dofs/dof_tools.cc [hunk 1]
```diff
@@ -227,9 +227,7 @@ namespace DoFTools
 
       // then loop over all cells and do the work
       std::vector<types::global_dof_index> indices;
-      for (typename DoFHandlerType::active_cell_iterator c = dof.begin_active();
-           c != dof.end();
-           ++c)
+      for (const auto c : dof.active_cell_iterators())
```
Comment (drwells) [inline source/dofs/dof_tools.cc]:
```suggestion
      for (const auto &c : dof.active_cell_iterators())
```

File: source/dofs/dof_tools.cc [hunk 2]
```diff
@@ -443,6 +441,70 @@ namespace DoFTools
...
+          this_selected_dofs[i] =
+            component_mask[local_component_association[i]];
+
+        local_selected_dofs.emplace_back(std::move(this_selected_dofs));
+      }
+
+    // Then loop over all cells and do the work
+    IndexSet                             selected_dofs(dof.n_dofs());
+    std::vector<types::global_dof_index> local_dof_indices;
+    for (const auto c : dof.active_cell_iterators())
```
Comment (drwells) [inline source/dofs/dof_tools.cc]:
```suggestion
    for (const auto &c : dof.active_cell_iterators())
```

**Comment by bangerth** [comment]
Thanks, @drwells. I've updated the code based on your suggestions.

**Pull request merged**